### PR TITLE
Add tfrecord format reader and tfrecord Example, Features protobuf generated classes

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

1. Add reader to read tfrecord format file
2. Add tfrecord Example, Features protobuf generated classes for ParseExample operator.
3. Upgrade protobuf version to 3.4.0 (manually tested on inception module loader and saver)

## How was this patch tested?

add unit tests

### CODE REVIEW

**Comment by jason-dai** [comment]
What do you mean by tf Example?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TFRecordIterator.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * Internal use only.
+ *
+ * TF record format:
+ *  uint64 length
+ *  uint32 masked_crc32_of_length
+ *  byte   data[length]
+ *  uint32 masked_crc32_of_data
+ *
+ */
+private[tf] class TFRecordIterator(fileName: File) extends Iterator[Array[Byte]] {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TFRecordIterator.scala]: this is local file only?

**Comment by yangw1234** [comment]
@jason-dai Example is a protobuf class provided by tensorflow to represent a record in tfrecod file. The tensorflow doc says  "The recommended format for TensorFlow is a TFRecords file containing tf.train.Example protocol buffers (which contain Features as a field)."

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TFRecordIterator.scala (hunk 1 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TFRecordIterator.scala]: Yes, for local file only.

**Comment by jason-dai** [comment]
So it's tfrecord example, yes?

**Comment by yangw1234** [comment]
pr validation pass

**Pull request merged**